elastic/kibana
# [Security Solutions] Adds telemetry stats for legacy notifications system

## Summary

Adds telemetry stats for the legacy notification system. Once enough users have naturally migrated from the legacy version to the non-legacy version we will remove the legacy notification system.

Adds the structure of:
```json
"legacy_notifications": {
  "total": 2
}
```

To the telemetry data underneath `detection_rules` like so:

```json
"detection_rules": {
  "detection_rule_detail": [],
  "detection_rule_usage": {
    "query": {
      "enabled": 1,
      "disabled": 1,
      "alerts": 391,
      "cases": 0
    },
    "threshold": {
      "enabled": 0,
      "disabled": 0,
      "alerts": 0,
      "cases": 0
    },
    "eql": {
      "enabled": 0,
      "disabled": 0,
      "alerts": 0,
      "cases": 0
    },
    "machine_learning": {
      "enabled": 0,
      "disabled": 0,
      "alerts": 0,
      "cases": 0
    },
    "threat_match": {
      "enabled": 0,
      "disabled": 0,
      "alerts": 0,
      "cases": 0
    },
    "elastic_total": {
      "enabled": 0,
      "disabled": 0,
      "alerts": 0,
      "cases": 0
    },
    "legacy_notifications": {
      "total": 2
    },
    "custom_total": {
      "enabled": 1,
      "disabled": 1,
      "alerts": 391,
      "cases": 0
    }
  }
}
```

* Pulls `getInitialDetectionMetrics` from e2e tests into the `security_solution/server/usage` for use in both e2e and unit tests.
* Simplifies unit test boiler plate by using `getInitialDetectionMetrics`
* Fixes `alert_version` missing within the types
* Fixes `rule_version` to convert from `string` to `number` for types to work with TypeScript. This also will fix the `rule_version` to be a number in the `_source` document for telemetry.
* Removes the `uuid` in the unit tests to keep things deterministic with just hard coded values. 

**Manual Testing/Code review checks**:

Good things to test are:
* This works across multiple spaces and not just default.
* This works with more than 1 legacy rule.
* The number drops as the legacy rules are upgraded by editing the rule and saving it.

To see telemetry go to advanced settings -> Usage Data (and click cluster data):
<img width="2193" alt="tel" src="https://user-images.githubusercontent.com/1151048/148457445-8e8455f0-06b0-451e-af85-ec86ff00e12b.png">

You should note that initially the structure is `0` like so:
<img width="778" alt="Screen Shot 2022-01-06 at 3 14 53 PM" src="https://user-images.githubusercontent.com/1151048/148459828-3ca5c12f-8813-4aaf-89c9-82e09c01947f.png">

Next either you should load an older version of Kibana and manually create actions to get the legacy notifications firing such as using 7.15 and then upgrade to the latest or use these dev instructions below. Note if you are using these instructions, then this isn't as true of a manual test as using a real upgraded system.

Create a rule and activate it normally within security_solution:
<img width="1124" alt="Screen Shot 2022-01-06 at 3 18 07 PM" src="https://user-images.githubusercontent.com/1151048/148460310-61381130-15ba-48fc-8d31-3d8aec6f452e.png">

Do not add actions to the rule at this point as we will first exercise the older legacy system. However, you want at least one action configured such as a slack notification:
<img width="575" alt="slack" src="https://user-images.githubusercontent.com/1151048/148460357-8080b635-0f07-4e82-84aa-35b18c575101.png">

Within dev tools do a query for all your actions and grab one of the `_id` of them without their prefix:
```json
GET .kibana/_search
{
  "query": {
    "term": {
      "type": "action"
    }
  }
}
```

Mine was `_id : action:879e8ff0-1be1-11ec-a722-83da1c22a481`, so I will be copying the ID of `879e8ff0-1be1-11ec-a722-83da1c22a481`

Go to the file `detection_engine/scripts/legacy_notifications/one_action.json` and add this id to the file. Something like this:

```json
{
  "name": "Legacy notification with one action",
  "interval": "1m",  <--- You can use whatever you want. Real values are "1h", "1d", "1w". I use "1m" for testing purposes.
  "actions": [
    {
      "id": "879e8ff0-1be1-11ec-a722-83da1c22a481", <--- My action id
      "group": "default",
      "params": {
        "message": "Hourly\nRule {{context.rule.name}} generated {{state.signals_count}} alerts"
      },
      "actionTypeId": ".slack" <--- I am a slack action id type.
    }
  ]
}
```

Query for an alert you want to add manually add back a legacy notification to it. Such as:
```json
# See all your siem.signals alert types and choose one
GET .kibana/_search
{
  "query": {
    "term": {
      "alert.alertTypeId": "siem.queryRule"
    }
  }
}
```

Grab the `_id` without the alert prefix. For mine this was `933ca720-1be1-11ec-a722-83da1c22a481`:

Within the directory of detection_engine/scripts execute the script:

```sh
./post_legacy_notification.sh 933ca720-1be1-11ec-a722-83da1c22a481
{
  "ok": "acknowledged"
}
```

which is going to do a few things. See the file detection_engine/routes/rules/legacy_create_legacy_notification.ts for the definition of the route and what it does in full, but we should notice that we have now have actions on the rule:
<img width="1050" alt="rule_actions" src="https://user-images.githubusercontent.com/1151048/148460761-016d57cd-2c1c-402d-a95e-853d0afa8414.png">

If you run telemetry now at this point you should see the number has increased like so:
<img width="782" alt="Screen Shot 2022-01-06 at 3 28 03 PM" src="https://user-images.githubusercontent.com/1151048/148461247-640c4ef9-b97d-49d7-8996-cd0a7943b0f5.png">

You should be able to repeat the steps to add any legacy notifications to different spaces and get the total count correctly across the spaces. You should also be able to migrate/update your rule and see the count decrease back to 0 eventually.

### Checklist
